Fix trait detection recursion for anonymous classes

herndlm opened the pull request:
Closes https://github.com/phpstan/phpstan/issues/6442

This was a missing edge case that lead to endless recursion in case the trait looking for and anonymous class using it were in the same file. By not skipping the node, it would start looking at the anonymous class, find the trait usage, start looking for the trait, find it, come to the anonymous class, find the trait usage, start looking for the trait, ..

Unfortunately I failed creating an AnalyzerIntegrationTest because of namespacing problems in the data file. Any help appreciated :)

ondrejmirtes commented:
Hi, I really need the test to be able to merge this. Can you push what you've tried to reproduce it?

herndlm commented:
Oh I have a different behaviour now on my other machine. Sure, I'll check that out and push that then. Might take a bit.

ondrejmirtes commented:
Also - please use the original source code (it has more classes): 

```php

trait T
{
    public function x(): void
	{
		\PHPStan\dumpType(parent::class);
	}
}

class A {}

class B extends A
{
	use T;
}

new class() extends B
{
	use T;
};

```

herndlm commented:
Very odd, it started behaving the same strange way now too. Initially it was working (and showing another recursion problem), then I added and removed classes from the test file and now I always get reflection errors. It feels like some caching or so could be the culprit, but even after deleting all caches I can think of I have the same errors :/
Let's see what CI tells..
I get e.g.
```
Reflection error: Bug6442\T not found
```

herndlm commented:
> Also - please use the original source code (it has more classes):
> 
> ```
> trait T
> {
>     public function x(): void
> 	{
> 		\PHPStan\dumpType(parent::class);
> 	}
> }
> 
> class A {}
> 
> class B extends A
> {
> 	use T;
> }
> 
> new class() extends B
> {
> 	use T;
> };
> ```

this was on purpose for now, as the PR here does not yet fix another recursion problem with `parent::` in the trait :/

ondrejmirtes commented:
So please fix all of the issues related to that bug report (otherwise it's not really a regression test for that bugreport and cannot be merged/closed). Thanks :)

ondrejmirtes commented:
Your reflection error might be fixable by running "composer dump" whenever you move/add classes.

herndlm commented:
> Your reflection error might be fixable by running "composer dump" whenever you move/add classes.

that was really it! thx! I was not thinking that composer might be doing anything with those test files. 
ok, I can now work on the full/clean fix :)

ondrejmirtes commented on tests/PHPStan/Analyser/AnalyserIntegrationTest.php:
```diff
@@ -513,6 +513,16 @@ public function testBug6253(): void
+	{
+		$errors = $this->runAnalyse(__DIR__ . '/data/bug-6442.php');
+		$this->assertCount(2, $errors);
+		$this->assertSame('Dumped type: \'Bug6442\\\A\'', $errors[0]->getMessage());
```
You can use `assertType()` instead and still assert no errors.

herndlm replied:
sorry, where would I do that? in the test file won't work I guess, because the same asserType would be called twice with different expected types  :/

ondrejmirtes replied:
Oh, alright!

herndlm commented:
well, looks like something is still/now broken on Windows

ondrejmirtes commented:
`$fileName === $originalClassFileName` will be false probably because of different directory separators

Use `FileHelper:normalizePath`.

herndlm commented:
maybe I should try to normalize it earlier / only once, give me a minute

ondrejmirtes commented:
Thank you!

herndlm commented:
Thank you too for the pointers 🎉
